docs/CODING_CONVENTIONS.md

```diff
@@ -0,0 +1,259 @@
+}
+```
+
+This pattern guards against any matches being non-exhaustive in case new error variants are added in the future.
+
+#### Use `amethyst_error::Error` for compositional APIs
+
+APIs which composes results from multiple crates _should_ use `amethyst_error::Error`.
+This is a generic error type which is capable of boxing any error and annotate it with debugging information.
+
+This must be used when defining APIs which composes errors generically, like with traits.
+This is also required when composing errors from other crates, since It is currently the only mechanism available to
+communicate backtraces.
+
+###### Do
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// no dependencies
+mod c {
+    use amethyst_error::Error;
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+###### Don't
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::A(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::B(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// depends on `a` and `b`.
+mod c {
+    use crate::{a, b};
+
+    pub enum Error {
+        A(a::error::Error),
+        B(b::error::Error),
+    }
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+#### Avoid overloading the default `Result`
+
+We _should not_ import a `Result` that overloads the default import.
+Instead, if we are using a `Result` alias, use it through a module (e.g. `io::Result`).
+
+This is a future-proofing pattern that prevents having to deal with conflicting `Result` types during refactoring or
+code re-use.
+Overloading `Result` also makes it harder to use the default `Result` when it's needed.
+This is especially useful when multiple modules export their own `Result` types.
+
+If a crate exports its own `Result` type, we _should_ use it through an `error` alias like this:
```

Comment:
I don't agree that we shouldn't override the std error inside the crate.

Reply:
Using a locally overloaded `Result` has generally been moved away from in the larger rust ecosystem in my experience.

I don't really mind it for small crates with no dependencies. But I'd rather err on the side of consistency across the project.

Reply:
I have the opposite impression. Also, if you don't want to overload the `Result` type, it makes much more sense to use `Result<T, MyError>` instead of `error::Result<T>` IMO. The former gives more type information and supports multiple error types.

Reply:
Updated based on this comment and a conversation we had in Discord. Does it look better now?